File: leetcode/linked-list/0002_add_two_numbers.py
# LC 2. Add Two Numbers | Medium
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val; self.next = next

class Solution:
    def addTwoNumbers(self, l1, l2):
        dummy = ListNode(); curr, carry = dummy, 0
        while l1 or l2 or carry:
            s = carry
            if l1: s += l1.val; l1 = l1.next
            if l2: s += l2.val; l2 = l2.next
            curr.next = ListNode(s % 10); carry = s // 10; curr = curr.next
        return dummy.next

File: leetcode/linked-list/0002_add_two_numbers.cpp
// LC 2. Add Two Numbers | Medium
struct ListNode {
    int val; ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummy(0); ListNode* cur = &dummy; int carry = 0;
        while (l1 || l2 || carry) {
            int s = carry;
            if (l1) { s += l1->val; l1 = l1->next; }
            if (l2) { s += l2->val; l2 = l2->next; }
            cur->next = new ListNode(s % 10); carry = s / 10; cur = cur->next;
        }
        return dummy.next;
    }
};
